program/src/lib.rs

```diff
@@ -37,3 +37,51 @@ pub fn get_violation_report_address(
         &id(),
     )
 }
+
+struct ViolationReportAddress<'a> {
+    address: Pubkey,
+    pubkey_seed: &'a [u8],
+    slot_seed: &'a [u8; 8],
+    violation_seed: [u8; 1],
+    bump_seed: [u8; 1],
+}
+
+impl<'a> ViolationReportAddress<'a> {
+    pub(crate) fn new(report: &'a ViolationReport) -> ViolationReportAddress<'a> {
+        let pubkey_seed = report.pubkey.as_ref();
+        let slot_seed = &report.slot.0;
+        let violation_seed = [report.violation_type];
+        let (pda, bump) =
+            Pubkey::find_program_address(&[pubkey_seed, slot_seed, &violation_seed], &id());
```

Comment:
The one annoying feature of this type is that it always derives the pda, which is a costly syscall, and it isn't used during `close`.

It might better to do the derivation in a separate function -- ie rename `key()` to `find_program_address()` and do the derivation there. What do you think?

Reply:
Turns out we don't need to use it in close. Also we do need to find the bump since it's part of the seeds so I think we'll need to have it for anyone that uses the seeds fn as well.